src/test_application/VehicleBuilder.cpp

```diff
@@ -24,223 +24,81 @@
 
-using namespace testapp;
-
-using osp::BlueprintPart;
-using osp::BlueprintMachine;
-using osp::PrototypePart;
-using osp::PCompMachine;
-using osp::DependRes;
-using osp::Vector3;
-using osp::Quaternion;
-using osp::machine_id_t;
-
-part_t VehicleBuilder::part_add(
-        DependRes<PrototypePart>& protoPart,
-        const Vector3& translation,
-        const Quaternion& rotation,
-        const Vector3& scale)
+namespace testapp
 {
-    uint32_t protoIndex = m_vehicle.m_prototypes.size();
-
-    // check if the Part Prototype is added already.
-    for (size_t i = 0; i < protoIndex; i ++)
-    {
-        const DependRes<PrototypePart>& dep = m_vehicle.m_prototypes[i];
 
-        if (dep == protoPart)
-        {
-            // prototype was already added to the list
-            protoIndex = i;
-            break;
-        }
-    }
+using osp::restypes::gc_importer;
 
-    if (protoIndex == m_vehicle.m_prototypes.size())
+VehicleBuilder::~VehicleBuilder()
+{
+    // clear resource owners
+    for (auto & [_, value] : m_prefabs)
     {
-        // Add the unlisted prototype to the end
-        m_vehicle.m_prototypes.emplace_back(protoPart);
+        m_pResources->owner_destroy(gc_importer, std::move(value.m_importer));
     }
 
-    size_t numMachines = protoPart->m_protoMachines.size();
-
-    // We now know which part prototype to initialize, create it
-    uint32_t blueprintIndex = m_vehicle.m_blueprints.size();
-    BlueprintPart &rPart = m_vehicle.m_blueprints.emplace_back(
-                protoIndex, numMachines, translation, rotation, scale);
-
-    // Add default machines from part prototypes
-
-    for (size_t i = 0; i < numMachines; i ++)
+    for (osp::PrefabPair &rPrefabPair : m_data.m_partPrefabs)
     {
-        PCompMachine const &protoMach = protoPart->m_protoMachines[i];
-        machine_id_t id = protoMach.m_type;
-        m_vehicle.m_machines.resize(std::max(m_vehicle.m_machines.size(),
-                                             size_t(id + 1)));
-
-        BlueprintMachine &rBlueprintMach = m_vehicle.m_machines[id].emplace_back();
-        rBlueprintMach.m_protoMachineIndex = i;
-        rBlueprintMach.m_partIndex = blueprintIndex;
-        rBlueprintMach.m_config = protoMach.m_config;
+        m_pResources->owner_destroy(gc_importer, std::move(rPrefabPair.m_importer));
     }
-
-    return part_t(blueprintIndex);
 }
 
-Vector3 VehicleBuilder::part_offset(
-        PrototypePart const& attachTo, std::string_view attachToName,
-        PrototypePart const& toAttach, std::string_view toAttachName) noexcept
+void VehicleBuilder::set_prefabs(std::initializer_list<SetPrefab> setPrefabs)
 {
-    Vector3 oset1{0.0f};
-    Vector3 oset2{0.0f};
-
-    for (osp::PCompName const& name : attachTo.m_partName)
+    for (SetPrefab const& set : setPrefabs)
     {
-        if (name.m_name == attachToName)
+        auto const foundIt = m_prefabs.find(set.m_prefabName);
+        if (foundIt != std::end(m_prefabs))
```

Comment:
pre-bind the result of std::end(m_prefabs) priot to the for-loop, unless you think you'll cause iterator invalidation.